### PR TITLE
Testing Environment comparing models in JSON format

### PR DESCRIPTION
**Pull request by juanlanuzag**
**WARNING: this branch should not be merged until feature/cadmium-dynamic is merged**

Testing Environment built using Boost.Test and accessing model structure through DEVDiagrammer generated JSON and Boost class property_tree.

For now the tests made for LI model only run for the cadmium dynamic model but I think this could be generalized easily (assuming you can make the JSON).

Tests end up getting kind of messy because you have to access all the model tree using property_tree, but I could find a better way to do it. Let me know if you have any ideas.

### CODE REVIEW

**Comment by sdavtaker** [comment]
Can you pull the last version of the branch and resolve the conflict mentioned by github and repost the PR?

**Comment by juanlanuzag** [comment]
> Can you pull the last version of the branch and resolve the conflict mentioned by github and repost the PR?

Done

**Comment by sdavtaker** [comment]
Thanks

File: src/dynamic/LI_generator.cpp [hunk 1]
```diff
@@ -79,7 +79,7 @@ std::shared_ptr<cadmium::dynamic::modeling::coupled<TIME>> create_LI_model(uint
         std::vector<std::shared_ptr<cadmium::dynamic::modeling::model>> atomics_current_level;
         if (level < depth) {
             //Last level does not have atomics
-            for(int idx_atomic=0; idx_atomic < width; idx_atomic++) {
+            for(int idx_atomic=0; idx_atomic < width-1; idx_atomic++) {
```
Comment (sdavtaker) [inline src/dynamic/LI_generator.cpp]: I would extract that "width -1" as  variable/constant and put a comment of why it is -1. We already missed-by-one error on it, lets leave a warning for future generations :)

File: test/cadmium_dynamic_LI_test.cpp [hunk 2]
```diff
@@ -0,0 +1,186 @@
+/**
+ * Copyright (c) 2013-2016, Damian Vicino
```
Comment (sdavtaker) [inline test/cadmium_dynamic_LI_test.cpp]: Did I?

File: test/cadmium_dynamic_LI_test.cpp [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+#include <boost/test/data/monomorphic.hpp>
+
+#include "test_helpers.hpp"
+#include "../src/dynamic/LI_generator.cpp"
+#include "../../DEVSDiagrammer/model_json_exporter/include/dynamic_json_exporter.hpp"
+
+namespace pt = boost::property_tree;
+namespace bdata = boost::unit_test::data;
+
+BOOST_AUTO_TEST_SUITE( cadmium_dynamic_LI_test_suite)
```
Comment (sdavtaker) [inline test/cadmium_dynamic_LI_test.cpp]: My idea was to output the model to a json file, have a directory having a list of "expected" json files, and a script that checks all generated files are "equal" to those expected.
Part of the code below can be used for a small command line binary comparing 2 files and telling they are "json" equal enough, or maybe we can jut relay on python or other language already supporting the comparison. 
However, if that below works, it already seems why more flexible and simple than what we had before.

**Comment by sdavtaker** [comment]
Squashed and merge, please pull before next PR.

File: test/cadmium_dynamic_LI_test.cpp (hunk 3 above)
Comment (juanlanuzag) [reply test/cadmium_dynamic_LI_test.cpp]: Making the tests in python starting with the json files sound better, the way you have to access the json using the property_tree isn't very good.

On the other hand, i think checking properties on the json structure is better than checking equallity. Specially thinking about having the same tests to run for multiple simulated versions; at least for the two cadmium versions, the json is no exactly the same (names of models, we actually read the lists in the json as sets, etc).

**Pull request merged**